### PR TITLE
feat(Carousel): improve styles inside GridLayout columns

### PR DESCRIPTION
**Pull request by wyunreal**
issue: https://github.com/Telefonica/mistica-web/issues/546

pending:
tests for the new hook,
tests for the carousels (I've changed it and no test failed, so I think I need to add tests).

Some result screenshots:

Small desktop:
carousel-on-small-desktop

Largue desktop:
carousels-on-largue-desktop

Mobile:
image

Tablet:
image

### CODE REVIEW

File: src/carousel.tsx [hunk 1]
```diff
@@ -101,6 +101,15 @@ const arrowButtonStyle = (theme: Theme) => ({
...
 
+const arrowButtonSeparation = (containerType: ContainerType, isLargueDesktop: boolean) => {
+    switch (containerType) {
+        case 'desktop-wide-column':
+            return isLargueDesktop ? -(24 + arrowButtonSize) : -arrowButtonSize / 2;
+        default:
+            return -arrowButtonSize / 2;
+    }
+};
+
```
Comment (wyunreal) [inline src/carousel.tsx:112]: @yceballost we need your opinion here, as this is not yet defined by design rules, this can change depending on your opinion (see screenshots attached).

File: src/grid-layout.tsx [hunk 2]
```diff
@@ -1,4 +1,5 @@
 import * as React from 'react';
+import ContainerSizeContext from './container-type-context';
```
Comment (wyunreal) [inline src/grid-layout.tsx:2]: @yceballost we need definition on which size (small, mid, wide should be each column on the template, I assumed small from 1 to 5, mid from 6 to 9 and wide just 10 and when not inside a grid layout.

Comment (yceballost) [reply src/grid-layout.tsx:2]: we can go with this by the moment and we will review when we can work on it

**Review by yceballost — CHANGES_REQUESTED** [review]

File: flow-defs/screen-size-context.js.flow [hunk 3]
```diff
@@ -7,6 +7,7 @@ export type ScreenSizeContextType = {
   isTabletOrBigger: boolean,
   isTabletOrSmaller: boolean,
   isDesktopOrBigger: boolean,
+  isLargueDesktop: boolean,
```
Comment (Winde) [inline flow-defs/screen-size-context.js.flow:10]: typo in Large

File: src/hooks.tsx [hunk 4]
```diff
@@ -185,3 +187,14 @@ export const useIsInViewport = (
 
     return isInViewport;
 };
+
+export const useContainerType = (): ContainerType => {
+    const containerType = React.useContext(ContainerTypeContext);
+    if (!containerType) {
+        throw Error(
```
Comment (Winde) [inline src/hooks.tsx:194]: Are we 100% sure that we want this to throw an error, there's no possibility of using one of the mistica atoms that use containerType without having this context?

File: src/responsive-layout.tsx [hunk 5]
```diff
@@ -64,16 +65,20 @@ const ResponsiveLayout: React.FC<Props> = ({children, className, fullWidth, data
 
     const sideMargin = isMobile ? MOBILE_SIDE_MARGIN : isTablet ? TABLET_SIDE_MARGIN : null;
 
+    const containerType = isMobile ? 'mobile-column' : isTablet ? 'tablet-column' : 'desktop-wide-column';
```
Comment (Winde) [inline src/responsive-layout.tsx:68]: hard to read ternary inside other ternary. 

I'd suggest to reword this to not chain them

File: src/screen-size-context-provider.tsx [hunk 6]
```diff
@@ -27,8 +27,15 @@ const ScreenSizeContextProvider: React.FC<Props> = ({children}) => {
...
             tabletOrSmaller: stripMedia(theme.mq.tabletOrSmaller),
+            largueDesktop: stripMedia(theme.mq.largeDesktop),
         }),
-        [theme.mq.mobile, theme.mq.tablet, theme.mq.tabletOrBigger, theme.mq.tabletOrSmaller]
+        [
+            theme.mq.mobile,
+            theme.mq.tablet,
+            theme.mq.tabletOrBigger,
+            theme.mq.tabletOrSmaller,
+            theme.mq.largeDesktop,
```
Comment (Winde) [inline src/screen-size-context-provider.tsx:37]: is this a new mediaquery? shouldn't it be a new definition part of this PR?

**Review by Winde — APPROVED** [review]

File: src/hooks.tsx (hunk 4 above)
Comment (wyunreal) [reply src/hooks.tsx:194]: mmm, the error is raised only when the hook is used on a component not placed on a responsive layout, nor a grid layout. By now, only the carousels are using it.

Question @yceballost: if not in a responsive layout nor grid layout, what value should arrive to the component via this hook ? maybe depending on the screen media queries ? I can do that if you think thats the way to go:
- mobile for mobile screen
- tablet for tablet
- desktop and wide desktop for desktop and large desktop respectively. 

WDYT ?

File: src/responsive-layout.tsx (hunk 5 above)
Comment (wyunreal) [reply src/responsive-layout.tsx:68]: I just copy pasted the pattern being seen on the above line ;) but you're right, its difficult to read.

File: src/screen-size-context-provider.tsx (hunk 6 above)
Comment (wyunreal) [reply src/screen-size-context-provider.tsx:37]: No no, the mediaquery already exists, I'm just adding it to this hook to make it available to components, as the carousel requires it.

File: src/hooks.tsx (hunk 4 above)
Comment (yceballost) [reply src/hooks.tsx]: for me it is ok!

In this way, the actual behavior keeps the same behavior, only changing when it is on a responsive or grid layout and the container is less than the mq defined, right?

**Review by yceballost — APPROVED** [review]
After some test, i didn't see anything weird! Thanks W <3

File: src/screen-size-context-provider.tsx [hunk 7]
```diff
@@ -27,8 +27,15 @@ const ScreenSizeContextProvider: React.FC<Props> = ({children}) => {
             tablet: stripMedia(theme.mq.tablet),
             tabletOrBigger: stripMedia(theme.mq.tabletOrBigger),
             tabletOrSmaller: stripMedia(theme.mq.tabletOrSmaller),
+            largueDesktop: stripMedia(theme.mq.largeDesktop),
```
Comment (pladaria) [inline src/screen-size-context-provider.tsx:30]: typo: largue

File: src/carousel.tsx [hunk 8]
```diff
@@ -114,22 +128,12 @@ const useStyles = createUseStyles((theme) => ({
         zIndex: 1,
         top: `calc(50% - ${arrowButtonSize / 2}px)`,
         '&.prev': {
-            left: -arrowButtonSize / 2,
-            [theme.mq.tabletOrSmaller]: {
```
Comment (atabel) [inline src/carousel.tsx:118]: I think I'd keep mediaqueries and just apply overrides in desktop when inside medium/small a grid column

File: src/responsive-layout.tsx [hunk 9]
```diff
@@ -64,16 +65,20 @@ const ResponsiveLayout: React.FC<Props> = ({children, className, fullWidth, data
...
+
     return (
         <ResponsiveLayoutMarginContext.Provider value={sideMargin}>
-            <div
-                className={classnames(classes.container, className)}
-                {...getPrefixedDataAttributes(dataAttributes)}
-            >
-                <div className={classnames(classes.layout, {[classes.layoutFullWidth]: fullWidth})}>
-                    {children}
+            <ContainerTypeContext.Provider value={containerType}>
```
Comment (atabel) [inline src/responsive-layout.tsx:72]: I wouldn't apply this to responsive layout, only to grid layout

File: src/container-type-context.tsx [hunk 10]
```diff
@@ -0,0 +1,27 @@
+import * as React from 'react';
+import {useScreenSize} from './hooks';
+
+import type {ContainerType} from './utils/types';
+
+const ContainerTypeContext = React.createContext<ContainerType | null>(null);
```
Comment (atabel) [inline src/container-type-context.tsx:6]: I'd call it `ContainerWidthContext` with values `'large' | 'medium' | 'small'` and only listen for them in desktop viewports. I won't mix screen size with container width

File: src/container-type-context.tsx [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+export const useContainerType = (): ContainerType => {
+    const containerType = React.useContext(ContainerTypeContext);
+    const {isTablet, isDesktopOrBigger, isLargeDesktop} = useScreenSize();
```
Comment (atabel) [inline src/container-type-context.tsx:10]: I won't mix screen size with container width, see my previous comment

File: src/utils/types.tsx [hunk 12]
```diff
@@ -10,3 +10,10 @@ export type IconProps = {
     className?: string;
     style?: React.CSSProperties;
 };
+
+export type ContainerType =
+    | 'desktop-small-column'
+    | 'desktop-mid-column'
```
Comment (pladaria) [inline src/utils/types.tsx:16]: "medium" instead of "mid" ?

File: src/utils/types.tsx [hunk 13]
```diff
@@ -10,3 +10,10 @@ export type IconProps = {
...
+    | 'desktop-wide-column'
```
Comment (pladaria) [inline src/utils/types.tsx:17]: "large" instead of "wide" ?

**Review by pladaria — APPROVED** [review]

**Comment by yceballost** [comment]
> image @yceballost @wyunreal, is this expected?

yes, talked with @wyunreal and we will improve when we can work on it later. Wiljan promised me that he will make the changes :P

File: src/carousel.tsx [hunk 14]
```diff
@@ -101,6 +102,19 @@ const arrowButtonStyle = (theme: Theme) => ({
     },
 });
 
+const arrowButtonSeparation = (containerType: ContainerType, isLargeDesktop: boolean, sideMargin: number) => {
```
[the remark this answers is not in the document]
Comment (wyunreal) [reply src/carousel.tsx:105]: Yes, I've decided to leave the balls the same (depending on the screen size) to let them feel the same (It would be weird to see some balls with different styles at the same time.
IMHO, we should not change styles of the bullets depending on the column.

File: src/carousel.tsx [hunk 15]
```diff
@@ -101,6 +102,19 @@ const arrowButtonStyle = (theme: Theme) => ({
...
+    console.log(containerType);
```
[the remark this answers is not in the document]
Comment (wyunreal) [reply src/carousel.tsx]: yes !! removed !!

File: src/container-type-context.tsx (hunk 10 above)
Comment (wyunreal) [reply src/container-type-context.tsx:6]: And what about the grid layout when its displayed in a mobile or tablet ? In these cases, the expected behavior is to be displayed as if the grid won't exists.

File: src/container-type-context.tsx (hunk 11 above)
Comment (wyunreal) [reply src/container-type-context.tsx:10]: I've answered, I agree we should not mix, thats why I'm not mixing at all XD, I've created the Columns alike enum to just denote a different concept: not the same (conceptually) a mobile screen than a mobile column

File: src/responsive-layout.tsx (hunk 9 above)
Comment (wyunreal) [reply src/responsive-layout.tsx:72]: Appliyng to responsive layout we can also enable the components to use the Context on all devices, simplifying the development of the components (they can then take into account only the container type).

File: src/screen-size-context-provider.tsx (hunk 7 above)
Comment (wyunreal) [reply src/screen-size-context-provider.tsx:30]: changed, thanks ;)

File: src/utils/types.tsx (hunk 12 above)
Comment (wyunreal) [reply src/utils/types.tsx]: yes, will change !!

**Review by atabel — APPROVED** [review]
